cache/redis_client.go

```diff
@@ -0,0 +1,347 @@
+
+	f.BoolVar(&c.TLSEnabled, prefix+".tls-enabled", false, "Enable connecting to Redis with TLS.")
+	c.TLS.RegisterFlagsWithPrefix(prefix, f)
+}
+
+func (c *RedisClientConfig) Validate() error {
+	if c.Endpoint.String() == "" {
+		return errRedisConfigNoEndpoint
+	}
+	// Set async only available when MaxAsyncConcurrency > 0.
+	if c.MaxAsyncConcurrency <= 0 {
+		return errRedisMaxAsyncConcurrencyNotPositive
+	}
+	return nil
+}
+
+type redisClient struct {
+	redis.UniversalClient
+	config RedisClientConfig
+
+	// Queue used to enqueue async operations.
+	asyncQueue *asyncQueue
+
+	// getMultiGate used to enforce the max number of concurrent GetMulti() operations.
+	getMultiGate gate.Gate
+
+	logger  log.Logger
+	metrics *clientMetrics
+}
+
+// NewRedisClient makes a new RedisClient.
+func NewRedisClient(logger log.Logger, name string, config RedisClientConfig, reg prometheus.Registerer) (RemoteCacheClient, error) {
+	opts := &redis.UniversalOptions{
+		Addrs:        strings.Split(config.Endpoint.String(), ","),
+		Username:     config.Username,
+		Password:     config.Password.String(),
+		DB:           config.DB,
+		MasterName:   config.MasterName,
+		DialTimeout:  config.DialTimeout,
+		ReadTimeout:  config.ReadTimeout,
+		WriteTimeout: config.WriteTimeout,
+		PoolSize:     config.ConnectionPoolSize,
+		MinIdleConns: config.MinIdleConnections,
+		MaxConnAge:   config.MaxConnectionAge,
+		IdleTimeout:  config.IdleTimeout,
+	}
+
+	if config.TLSEnabled {
+		tlsConfig := config.TLS
+
+		tlsClientConfig, err := newClientConfig(logger, tlsConfig.CertPath, tlsConfig.KeyPath,
+			tlsConfig.CAPath, tlsConfig.ServerName, tlsConfig.InsecureSkipVerify)
+
+		if err != nil {
+			return nil, err
+		}
+
+		opts.TLSConfig = tlsClientConfig
+	}
+
+	cli := redis.NewUniversalClient(opts)
+	if reg != nil {
+		reg = prometheus.WrapRegistererWith(prometheus.Labels{"name": name}, reg)
+	}
+
+	c := &redisClient{
+		UniversalClient: cli,
+		config:          config,
+		logger:          logger,
+	}
+	if config.MaxGetMultiConcurrency > 0 {
+		c.getMultiGate = gate.New(
+			prometheus.WrapRegistererWithPrefix("redis_getmulti_", reg),
+			config.MaxGetMultiConcurrency,
+		)
+	}
+
+	c.metrics = newClientMetrics(
+		prometheus.WrapRegistererWithPrefix("redis_", reg),
+	)
+
+	c.asyncQueue = newAsyncQueue(config.MaxAsyncBufferSize, config.MaxAsyncConcurrency)
+
+	return c, nil
+}
+
+// SetAsync implement RemoteCacheClient.
+func (c *redisClient) SetAsync(ctx context.Context, key string, value []byte, ttl time.Duration) error {
+	// Skip hitting redis at all if the item is bigger than the max allowed size.
+	if c.config.MaxItemSize > 0 && uint64(len(value)) > uint64(c.config.MaxItemSize) {
+		c.metrics.skipped.WithLabelValues(opSet, reasonMaxItemSize).Inc()
+		return nil
+	}
+
+	err := c.asyncQueue.submit(func() {
+		start := time.Now()
+
```

Comment:
Shouldn't we do `c.metrics.operations.WithLabelValues(opSet).Inc()` here as well, like we do in the memcached client?